api/cas/login.py

```diff
@@ -0,0 +1,216 @@
+    if verification_key:
+        if verification_key == user.verification_key:
+            return user, verify_two_factor(user, one_time_password), util.is_user_inactive(user)
+        return None, errors.INVALID_VERIFICATION_KEY, None
+
+    # authenticated by password
+    if password:
+        if user.check_password(password):
+            return user, verify_two_factor(user, one_time_password), util.is_user_inactive(user)
+        return None, errors.INVALID_PASSWORD, None
+
+
+def handle_login_institution(provider):
+    """
+    Handle login through institution identity provider.
+
+    :param provider: the decrypted provider object
+    :return: the user
+    :raises: ValidationError
+    """
+
+    # check required fields
+    institution_id = provider.get('id', '')
+    institution_user = provider.get('user', '')
+    if not (institution_id and institution_user):
+        raise ValidationError(detail=messages.INVALID_REQUEST)
+
+    # check institution
+    institution = Institution.load(institution_id)
+    if not institution:
+        raise ValidationError(detail='Invalid institution id specified "{}"'.format(institution_id))
+
+    # check username (email)
+    username = institution_user.get('username', '')
+    if not username:
+        message = 'Institution login failed: username required'
+        sentry.log_message(message)
+        raise ValidationError(detail='Institution login failed: username required')
+
+    # check fullname
+    fullname = institution_user.get('fullname', '')
+    given_name = institution_user.get('givenName', '')
+    family_name = institution_user.get('familyName', '')
+    if not fullname:
+        fullname = '{} {}'.format(given_name, family_name).strip()
+    if not fullname:
+        message = 'Institution login failed: fullname required for user {} from institution {}'\
+            .format(username, institution_id)
+        sentry.log_message(message)
+        raise ValidationError(detail=message)
+
+    middle_names = institution_user.get('middleNames')
+    suffix = institution_user.get('suffix')
+
+    # get or create user
+    user, created = get_or_create_user(fullname, username, reset_password=False)
+
+    if created:
+        # replace the guessed names
+        user.given_name = given_name if given_name else user.given_name
+        user.family_name = family_name if family_name else user.family_name
+        user.middle_names = middle_names if middle_names else user.middle_names
+        user.suffix = suffix if suffix else user.suffix
+        user.update_date_last_login()
+        # save and register user
+        user.save()
+        user.register(username)
+        # send confirmation email
+        send_mail(to_addr=user.username, mail=WELCOME_OSF4I, mimetype='html', user=user)
+
+    if not user.is_affiliated_with_institution(institution):
+        user.affiliated_institutions.add(institution)
+        user.save()
+
+    return user, None
+
+
+def handle_login_external(data_user):
```

Comment:
New! This was handled as a hack during CAS-OSF service ticket exchange: https://github.com/CenterForOpenScience/osf.io/blob/develop/framework/auth/cas.py#L247. The next comment in this review shows the removal of that specific hack.